prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,238 @@
+		withTLS:     tls,
+	}
+}
+
+// NewGCSJobSource creates a new GCSJobSource from a given bucket and jobPath
+func NewGCSJobSource(src string) *GCSJobSource {
+	linkPrefix := "gs://"
+	noPrefixSrc := strings.TrimPrefix(src, linkPrefix)
+	if !strings.HasSuffix(noPrefixSrc, "/") { // Cleaning up path
+		noPrefixSrc += "/"
+	}
+	tokens := strings.FieldsFunc(noPrefixSrc, func(c rune) bool { return c == '/' })
+	bucket := tokens[0]
+	buildID := tokens[len(tokens)-1]
+	name := tokens[len(tokens)-2]
+	jobPath := strings.TrimPrefix(noPrefixSrc, bucket+"/") // Extra / is not part of prefix, only necessary for URI
+	return &GCSJobSource{
+		source:     src,
+		linkPrefix: linkPrefix,
+		bucket:     bucket,
+		jobPath:    jobPath,
+		jobName:    name,
+		buildID:    buildID,
+	}
+}
+
+// isGCSSource recognizes whether a source string references a GCS bucket
+func isGCSSource(src string) bool {
+	return strings.HasPrefix(src, "gs://")
+}
+
+// GCSMarker holds the starting point for the next paginated GCS query
+type GCSMarker struct {
+	XMLName xml.Name `xml:"ListBucketResult"`
+	Marker  string   `xml:"NextMarker"`
+}
+
+// Contents is a single entry returned by the GCS XML API
+type Contents struct {
+	Key string
+}
+
+// GCSReq contains the contents of a GCS XML API list response
+type GCSReq struct {
+	XMLName  xml.Name `xml:"ListBucketResult"`
+	Contents []Contents
+}
+
+// names is a helper function for extracting artifact names in parallel from a GCS XML API response
+func names(content []byte, names chan string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	extracted := GCSReq{
+		Contents: []Contents{},
+	}
+	err := xml.Unmarshal(content, &extracted)
+	if err != nil {
+		logrus.WithError(err).Error("Error unmarshaling artifact names from XML")
+	}
+	for _, c := range extracted.Contents {
+		names <- c.Key
+	}
+}
+
+// Artifacts lists all artifacts available for the given job source.
+// Uses the GCS XML API because it is ~2x faster than the golang GCS library
+// for large number of artifacts. It should also be S3 compatible according to
+// the GCS api docs.
+// TODO: Look at pushing this upstream to golang GCS client
+func (af *GCSArtifactFetcher) Artifacts(src JobSource) []string {
+	var wg sync.WaitGroup
+	artStart := time.Now()
+	artifacts := []string{}
+	prefix := src.JobPath()
+	maxResults := 1000
+	bodies := [][]byte{}
+	marker := GCSMarker{}
+	c := http.Client{}
+	scheme := "https"
+	if !af.withTLS {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		c = http.Client{Transport: tr}
+		scheme = "http"
+	}
+	for {
+		params := url.Values{}
+		params.Add("prefix", prefix)
+		if marker.Marker != "" {
+			params.Add("marker", marker.Marker)
+		}
+		req := url.URL{
+			Scheme:   scheme,
+			Host:     af.xmlEndpoint,
+			Path:     src.BucketName(),
+			RawQuery: params.Encode(),
+		}
+		resp, err := c.Get(req.String())
+		if err != nil {
+			logrus.WithError(err).Error("Error in GCS XML API GET request")
```

Comment:
hummm, break here? What's the expect behavior if the request didn't get through?

Reply:
added retries, broke into separate function

Reply:
On the log side of things -- you want to add fields to logs like this to identify the request that was being made